src/coreclr/jit/morph.cpp

```diff
@@ -13892,13 +13891,70 @@ PhaseStatus Compiler::fgMorphBlocks()
 
     // Morph all blocks.
     //
-    // Note morph can add blocks downstream from the current block,
-    // and alter (but not null out) the current block's bbNext;
-    // this iterator ensures they all get visited.
-    //
-    for (BasicBlock* block : Blocks())
+    if (!optLocalAssertionProp)
     {
-        fgMorphBlock(block);
+        // If we aren't optimizing, we just morph in normal bbNext order.
+        //
+        // Note morph can add blocks downstream from the current block,
+        // and alter (but not null out) the current block's bbNext;
+        // this iterator ensures they all get visited.
+        //
+        for (BasicBlock* block : Blocks())
+        {
+            fgMorphBlock(block);
+        }
+    }
+    else
+    {
+        // We are optimizing. Process in RPO.
+        //
+        fgRenumberBlocks();
+        EnsureBasicBlockEpoch();
+        fgComputeEnterBlocksSet();
```

Comment:
How much of the cost comes from having to do the above setup calls before even computing the order?
When I've tried to compute the order before it wasn't clear to me why I would need to renumber blocks to do that.

Reply:
I can grab a new profile, but the old one is probably pretty accurate: see https://github.com/dotnet/runtime/pull/86822#issuecomment-1626319133

Renumbering likely isn't needed and not renumbering should save a bit of TP. So let me try that.

Also looking back at that early PR reminded me that I should take a look at if SSA's topological sort, it might be more efficient.

Reply:
It's also possible we can make `DfsBlockEntry` cheaper in the same way as #86839 (even though `GetSucc` looks cheap, it would probably still help a bit).

It also confuses me a bit that we have multiple notions of start nodes with `fgEnterBlks` and `fgDomFindStartNodes`. I assume the nodes found exclusively by `fgDomFindStartNodes` are unreachable (?) so ideally we would get rid of them earlier, but even if not, isn't it optimizable by using `bbRefs`? If not, then I think using `VisitSuccs` would be more efficient.

Anyway, don't feel the need to address it in this PR -- but might be a couple of avenues of future cleanup/improvements.

Reply:
Seems like for now at least renumbering is needed. Let me add some notes to #93246 about making all this more efficient. I'll come back to that after I get the assertion prop part working.